### ISSUE TITLE
[Question / Feature Request] AWS NLB - EIP Allocations

### ISSUE THREAD

**Issue by timothyclarke** [kubernetes/kops#10706]
Firstly thanks for the work people have put in for the the Network Loadbalancer in AWS clusters.

Is there a way (without using `-out ... --target=terraform` and editing the tf config) to use IP allocations so we can make use of static IP's?

I haven't thought through the cli elements to create the cluster, but I would expect that the spec would be something like
```
spec:
  api:
    loadBalancer:
      class: Network
      type: Public
      elasticIPs:
      - eipalloc-000abc123
      - eipalloc-111def456
      - eipalloc-222ghi789
```

**Comment by bharath-123** [comment]
I don't think kOps currently supports specifying allocationIds of elasticIps to attach to a network load balancer. You probably would have to manually associate an elastic ip at this point of time. 

But I do see a way to do this in kOps. The `SubnetMapping` struct in `CreateLoadBalancerInput` accepts an allocationId for an elasticIp for a specified subnet.([link](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/elasticloadbalancingv2@v1.1.0/types#SubnetMapping)) which can be used to associate elastic ips to the nlb.

**Comment by bharath-123** [comment]
I think if we can specify a mapping of subnetId to allocationId of the elastic ip in the lb spec, we should be able to associate elastic ips to the nlb.

**Comment by rifelpet** [comment]
Agreed, I think it would be reasonable to support this. We'll need the new API fields to be able to represent which subnet ID to use for each allocation ID. I _believe_ that the [name](https://github.com/kubernetes/kops/blob/9d0dd3c4c1957f829504376778d0ee67f8f4d28f/tests/integration/update_cluster/complex/in-v1alpha2.yaml#L68) field of subnets in the ClusterSpec are unique, so we might be able to get away with referencing them by name


Maybe something along the lines of:

```yaml
spec:
  api:
    loadBalancer:
      class: Network
      type: Public
      subnetMapping: # or a more generic term for this
      - subnet: us-east-1a-public
        allocationID: eipalloc-111def456
      - subnet: us-east-1b-public
        allocationID: eipalloc-222ghi789
```

Then we just find the subnet ID associated with each subnet name and pass all of that into the CreateLoadBalancerInput.

**Comment by timothyclarke** [comment]
Shouldn't the subnet ID's come (indirectly) from the master instance groups ?
I also noticed that the **Cross-zone load balancing** is disabled by default. Wouldn't that cause issues when one or more of the masters is down ?

**Comment by rifelpet** [comment]
Yes we can consider grabbing the subnets from the master instance groups and fetching their IDs. That would likely be more accurate than just the list of subnets in the ClusterSpec.

We can discuss enabling cross-zone load balancing by default for new clusters, apparently [it was suggested](https://github.com/kubernetes/kops/pull/6958#issuecomment-504252194) when it was first introduced :)

And we can add validation that enforces immutability for the proposed API fields.

**Comment by bharath-123** [comment]
@timothyclarke I tried to add an elastic ip to a pre-created load balancer but it looks like that is not possible as you said. AWS puts not being able to allocate an elastic ip to the nlb as a 'Temporary Limitation' for some reason. Based on this doc https://aws.amazon.com/premiumsupport/knowledge-center/elb-attach-elastic-ip-to-public-nlb/ I assumed that we could edit existing nlbs to allocated eips.

@rifelpet don't we want the users to be able choose which subnet he/she wants to allocated an elastic ip? What if a user only wants to allocate an elastic ip to 2 out of the 3 AZs in which the master is in? Do you think that is a valid use case?

**Comment by timothyclarke** [comment]
>What if a user only wants to allocate an elastic ip to 2 out of the 3 AZs in which the master is in? Do you think that is a valid use case?

Personally I don't think that is a valid use case. The reason I was watching this one (and am looking to implement) are due to remote working / WFH reasons.
* We do not want the admin interface to be accessible to world + dog so an IP restrictions are in place.
* As a business we do not want to have a VPN to every cluster for management purposes same goes for a vpn to a transit vpc which then peers to each cluster. 
  * Providing additional routing will have people starting to introduce `NodePort` connections where they should not be instead of using `kubectl port-forward`
* API access must come from known IP's, which equates to corporate VPN or home IP's (for those who have a static IP). 
* Classic ELB's have dynamic IP's so impose regular updates to VPN's to maintain correct routing rules. Sadly the VPN appliance we're using has IP based rather than host name based routing for clients on the vpn. 
* NLB's allow for fixed IP addresses and so removes the need for updating the routes on the VPN. 

Having some IP's that are static and some that are dynamic provide no logistical benefit. Assuming 3 IP's for the management interface connections to the api would fail 33% or 66% of the time when it attempts to use the dynamic IP address. We're talking about the external management interface so I think the use case I've described above would cover most situations. 

The only variance to the above would be if you decided to have `spec.api.loadBalancer.crossZoneLoadBalancing: true` and then a dns entry that only included the elastic ip's. Unless the spec is going to be extended to support this as a valid use case I think it should not be catered for.

**Comment by bharath-123** [comment]
@timothyclarke I get your points. I was looking at giving more flexibility to the user. I think you would be interested in this PR https://github.com/kubernetes/kops/pull/10741

This PR is doing exactly what I proposed i.e allowing the user to pass in the subnets and specify the elastic IP. But in this PR case, the author has only considered privateIPv4address for the case of an internal load balancer. 

Do have a look at it, I believe your input will be very valuable here given that your looking to implement this.

**Comment by timothyclarke** [comment]
@bharath-123 Thank you.

If I take this back to requirements, rather than implementation. 
* I'm after static IP's for the external management API. 
* I'm not at the stage where I want to output terraform and apply that, however I'm not averse to terraform and I perform the initial VPC setup via terraform.

KOPS already allows us to add additional security groups to the masters I'd be happy if I could switch off the external loadbalancer and create one independantly. eg
```
spec:
  api:
    loadBalancer:
      class :  DNSonly
      dnsType :  cname
      dnsTarget :  abc-123.elb.us-east-1.amazonaws.com
```
I've seen [a blog post](https://blog.scottlowe.org/2019/07/30/adding-a-name-to-kubernetes-api-server-certificate/) gives some details how to update the API cert SAN's (just not from within KOPS). So I'd be equally happy if I could set `spec.api.loadBalancer.type: Internal` and then add an external one following the same logic as above. The only issue I can see with the extra SAN is that I'd have to amend the output of `kops export kubecfg` to make use of the new hostname, or have an additional flag so it would provide a corrected output.

Sorry if this appears a little random & off topic. Perhaps if anyone with an opinion on this provide input. Is my usecase common enough that it should be supported by KOPS, or is static piblic (EIP) IP's for the K8s API and edge case that so few would use that it is not worth KOPS supporting and maintaining that feature.

**Comment by rifelpet** [comment]
One option you might consider exploring is the [externalLoadBalancers](https://pkg.go.dev/k8s.io/kops/pkg/apis/kops#InstanceGroupSpec) InstanceGroupSpec field. It allows you to define a classic load balancer or an ALB/NLB target group that will be attached to the ASG. You could provision an NLB and TargetGroup externally with terraform and then pass the TG ARN to each of the master instance groups. that should work in terms of providing kube-apiserver access.

The shortfalls of that approach is, as you mentioned `kops export kubecfg` would still use the api load balancer known to kops rather than the terraform one. Also the `UseForInternalApi` field that forces kubelets to connect to kube-apiserver via the load balancer would go through the kops load balancer rather than the terraform load balancer.

**Comment by timothyclarke** [comment]
@rifelpet externalLoadBalancers InstanceGroupSpec was more or less the route I was thinking (I'm doing similar for nginx ingress). My question however is then on the DNS / SAN side of things. 
While I can create the external loadbalancer the DNS name will not be in the SAN that the clusters SSL cert provides. `kops export kubecfg | sed` is not an insurmountable problem.
As far as I was aware the default internal (pod to master) usage is roundrobin DNS rather than via a loadbalancer

**Comment by rifelpet** [comment]
Yes what you said is correct, you'll likely run into cert issues and need to recreate the master instances.

I think that https://github.com/kubernetes/kops/pull/10741 will get us most of the way towards what you're after, and we can add the Allocation ID API field in a followup PR allowing you to use EIPs created outside of kops. Would that be sufficient for your usecase?

**Comment by timothyclarke** [comment]
Such changes would be sufficient in my use case. https://github.com/kubernetes/kops/pull/10741 looks good. 
Once it's in I'll probably be happy to attempt a PR of my own extending that code. I'm a little rusty on Go so hopefully any code I produce isn't toooo bad. 
The BLM logo's on the [contributing](https://kops.sigs.k8s.io/welcome/contributing/) and [building(https://kops.sigs.k8s.io/welcome/contributing/contributing/building/) pages make it a little harder to start / know how / where to start and how to build binaries locally for test prior to raising a PR

**Comment by timothyclarke** [comment]
Now that https://github.com/kubernetes/kops/pull/10741 has been merged I've taken a first pass at updating kops to make use of the ElasticIP's. The diff is at https://github.com/kubernetes/kops/compare/master...timothyclarke:NLB-EIP however I'm getting the following error when I try to edit the cluster to replace the Classic LP with a NLB with a EIP  (I've changed the allocation id, I am using a real one)
```
# Found fields that are not recognized
# ...
#         class: Network
#         subnets:
# +       - allocationId: eipalloc-00000000
# +         name: eu-west-1b
# -       - name: eu-west-1b
#         type: Public
#     authorization:
```

I'm not a Go developer so I've created a [PR within my own branch](https://github.com/timothyclarke/kops/pull/1) so I'm not flooding this thread with code comments. Any constructive criticism would be welcomed. I'm quite sure there's a bunch of tests I'll need to add, I was trying to get to a functional state.
I suspect that I might need to resolve the `eipalloc` to the actual IP (like how the NAT gateways are) rather than simply passing the allocationId to the NLB. If that is the case I'm highly likely to be out of my depth.

**Comment by timothyclarke** [comment]
PR has been merged and will be in 1.20.  closing
